tests/examples/wibox/widget/template/concrete_implementation.lua

```diff
@@ -0,0 +1,55 @@
+--DOC_NO_USAGE
+
+--DOC_HIDE_START
+local parent = ...
+
+local gears = require("gears")
+local wibox = require("wibox")
+
+local concrete_widget_template_builder
+--DOC_HIDE_END
+
+    -- Build the default widget used as a fallback if user doesn't provide a template
+    local default_widget = {
+        template = wibox.widget.textclock,
+        update_callback = function(widget_template, args)
+            local text = args.text or "???"
+            widget_template.widget.text = text
+        end,
+    }
+
+--DOC_NEWLINE
+    function concrete_widget_template_builder(args)
+        -- TODO : this should be made easier
+        local ret = args.widget_template
+                and args.widget_template.is_widget
+                and args.widget_template.widget_name == "wibox.widget.template"
+                and args.widget_template
+            or wibox.widget.template(
+                args.widget_template or default_widget
+            )
```

Comment:
You could create a static function `wibox.widget.template.ensure_is_template(widget_template, default_widget)` that does this.

Reply:
I have this other branch (unstable since I rebase it every time I update the template widget) where I implement the usage of this new widget in the scope of the layoutbox (https://github.com/Aire-One/awesome/tree/feature/widget_template/layoutbox).

I took this big "if test" from there. Having a static function that handle that is actually an improvement I thought of. (It was part of the idea behind the "proxy contractor" I thought of in my todo list from the first message.)

Another possible approach would be to make the widget constructor capable of determining you gave a working instance to return it directly, or clone it maybe (?) I'm not sure whether it would make sense, tho...

Another way I'd like to explore is to use the make_widget function with the template widget as a proxy (https://awesomewm.org/apidoc/classes/wibox.widget.base.html#wibox.widget.base.make_widget). I never used this function that way before, so I'm not sure how it would actually behave. I need to dig the source code on that...

Reply:
I went for the easy way : use a basic table with parameters to pass to the template, and if this table doesn't exist, fallback to the default template.

There is no easy way to enforce typing of parameters without overcomplicated checks that doesn't add much to the end user. (And the test I did there was flaky anyway, you just can't find the widget type based on the widget_name property, that's a bad idea.)